File: edu-ios/Sources/Views/DrillDiagram.swift
import SwiftUI

// EXPERIMENT — inline diagrams for the geometry-flavored drills, drawn with
// Canvas (no assets). A unit-circle dial with the angle's radius + the asked
// component highlighted, and a vector arrow with its x/y component called out.
struct DrillDiagram: View {
    let spec: DrillDiagramSpec
    var size: CGFloat = 168

    var body: some View {
        Canvas { ctx, sz in
            let c = CGPoint(x: sz.width / 2, y: sz.height / 2)
            let r = min(sz.width, sz.height) / 2 - 16
            switch spec {
            case let .unitCircle(deg, fn): drawUnitCircle(&ctx, c, r, deg, fn)
            case let .vector(deg, comp): drawVector(&ctx, c, r, deg, comp)
            }
        }
        .frame(width: size, height: size)
    }

    private func point(_ c: CGPoint, _ deg: Double, _ r: CGFloat) -> CGPoint {
        let a = deg * .pi / 180
        return CGPoint(x: c.x + r * CGFloat(cos(a)), y: c.y - r * CGFloat(sin(a))) // y up
    }

    private func line(_ ctx: inout GraphicsContext, _ p1: CGPoint, _ p2: CGPoint,
                      _ color: Color, _ w: CGFloat, dashed: Bool = false) {
        var path = Path()
        path.move(to: p1); path.addLine(to: p2)
        let style = dashed ? StrokeStyle(lineWidth: w, dash: [4, 3]) : StrokeStyle(lineWidth: w, lineCap: .round)
        ctx.stroke(path, with: .color(color), style: style)
    }

    private func axes(_ ctx: inout GraphicsContext, _ c: CGPoint, _ r: CGFloat) {
        line(&ctx, CGPoint(x: c.x - r - 8, y: c.y), CGPoint(x: c.x + r + 8, y: c.y), Theme.line, 1)
        line(&ctx, CGPoint(x: c.x, y: c.y - r - 8), CGPoint(x: c.x, y: c.y + r + 8), Theme.line, 1)
    }

    private func dot(_ ctx: inout GraphicsContext, _ p: CGPoint, _ color: Color, _ rad: CGFloat = 4) {
        ctx.fill(Path(ellipseIn: CGRect(x: p.x - rad, y: p.y - rad, width: rad * 2, height: rad * 2)), with: .color(color))
    }

    private func arrowhead(_ ctx: inout GraphicsContext, from: CGPoint, to: CGPoint, _ color: Color) {
        let ang = atan2(to.y - from.y, to.x - from.x)
        let len: CGFloat = 9, spread = CGFloat.pi / 7
        for s in [spread, -spread] {
            let p = CGPoint(x: to.x - len * cos(ang - s), y: to.y - len * sin(ang - s))
            line(&ctx, to, p, color, 2.5)
        }
    }

    private func drawUnitCircle(_ ctx: inout GraphicsContext, _ c: CGPoint, _ r: CGFloat, _ deg: Double, _ fn: String) {
        axes(&ctx, c, r)
        ctx.stroke(Path(ellipseIn: CGRect(x: c.x - r, y: c.y - r, width: r * 2, height: r * 2)),
                   with: .color(Theme.inkSoft.opacity(0.5)), lineWidth: 1.5)
        let p = point(c, deg, r)
        line(&ctx, c, p, Theme.ink, 2)                    // radius
        let accent = Theme.gold300
        if fn == "cos" {
            line(&ctx, c, CGPoint(x: p.x, y: c.y), accent, 3)          // horizontal component
            line(&ctx, p, CGPoint(x: p.x, y: c.y), accent.opacity(0.6), 1.5, dashed: true)
        } else if fn == "sin" {
            line(&ctx, c, CGPoint(x: c.x, y: p.y), accent, 3)          // vertical component
            line(&ctx, p, CGPoint(x: c.x, y: p.y), accent.opacity(0.6), 1.5, dashed: true)
        }
        dot(&ctx, p, accent, 4.5)
    }

    private func drawVector(_ ctx: inout GraphicsContext, _ c: CGPoint, _ r: CGFloat, _ deg: Double, _ comp: String) {
        axes(&ctx, c, r)
        let p = point(c, deg, r)
        line(&ctx, c, p, Theme.ink, 2.5)                  // the vector
        arrowhead(&ctx, from: c, to: p, Theme.ink)
        let accent = Theme.gold300
        if comp == "x" {
            line(&ctx, c, CGPoint(x: p.x, y: c.y), accent, 3.5)
            line(&ctx, p, CGPoint(x: p.x, y: c.y), accent.opacity(0.6), 1.5, dashed: true)
        } else {
            line(&ctx, c, CGPoint(x: c.x, y: p.y), accent, 3.5)
            line(&ctx, p, CGPoint(x: c.x, y: p.y), accent.opacity(0.6), 1.5, dashed: true)
        }
        dot(&ctx, p, accent, 4)
    }
}

File: edu-android/app/src/main/java/com/timpsonlyceum/lyceum/ui/screens/DrillDiagram.kt
package com.timpsonlyceum.lyceum.ui.screens

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.timpsonlyceum.lyceum.drills.DrillDiagram
import com.timpsonlyceum.lyceum.ui.theme.Theme
import com.timpsonlyceum.lyceum.ui.theme.serif
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/** The figure that goes with a drill prompt, when it has one. */
@Composable
fun DrillDiagramView(diagram: DrillDiagram) {
    when (diagram) {
        is DrillDiagram.Matrix -> BracketedGrid(diagram.rows, bars = true)
        is DrillDiagram.MatrixVector -> Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            BracketedGrid(diagram.matrix, bars = false, fill = false)
            Spacer(Modifier.width(10.dp))
            BracketedGrid(diagram.vector.map { listOf(it) }, bars = false, fill = false)
        }
        is DrillDiagram.UnitCircle -> UnitCirclePlot(diagram.angleDeg, diagram.fn)
        is DrillDiagram.Vector -> VectorPlot(diagram.angleDeg, diagram.component)
        is DrillDiagram.GeoMap -> Text(
            "Map not available",
            style = serif(14).copy(color = Theme.inkSoft),
        )
    }
}

/**
 * A matrix, drawn with determinant bars or square brackets.
 *
 * The minus signs are replaced here rather than at the generator, because the
 * grid is the only place these integers are shown raw.
 */
@Composable
private fun BracketedGrid(rows: List<List<Int>>, bars: Boolean, fill: Boolean = true) {
    val edge = if (bars) "│" else "["
    val edgeR = if (bars) "│" else "]"
    Row(
        modifier = if (fill) Modifier.fillMaxWidth() else Modifier,
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(edge, fontSize = (28 * rows.size).sp, color = Theme.gold400)
        Spacer(Modifier.width(6.dp))
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            rows.forEach { row ->
                Row {
                    row.forEach { v ->
                        Text(
                            v.toString().replace("-", "−"),
                            style = serif(20).copy(color = Theme.ink, textAlign = TextAlign.Center),
                            modifier = Modifier.widthIn(min = 44.dp).padding(vertical = 2.dp),
                        )
                    }
                }
            }
        }
        Spacer(Modifier.width(6.dp))
        Text(edgeR, fontSize = (28 * rows.size).sp, color = Theme.gold400)
    }
}

/** The unit circle with the asked-about angle drawn on it. */
@Composable
private fun UnitCirclePlot(angleDeg: Double, fn: String) {
    Canvas(Modifier.fillMaxWidth().height(180.dp)) {
        val cx = size.width / 2
        val cy = size.height / 2
        val r = minOf(cx, cy) * 0.78f

        drawCircle(Theme.line, radius = r, center = Offset(cx, cy), style = Stroke(width = 2f))
        drawLine(Theme.line, Offset(cx - r, cy), Offset(cx + r, cy), strokeWidth = 1.5f)
        drawLine(Theme.line, Offset(cx, cy - r), Offset(cx, cy + r), strokeWidth = 1.5f)

        // Screen y grows downward, so the angle is negated to read anticlockwise.
        val rad = angleDeg * PI / 180.0
        val px = cx + r * cos(rad).toFloat()
        val py = cy - r * sin(rad).toFloat()

        drawLine(Theme.gold300, Offset(cx, cy), Offset(px, py), strokeWidth = 4f)
        drawCircle(Theme.gold300, radius = 7f, center = Offset(px, py))

        // The dashed drop shows which projection the question is asking for.
        val dash = PathEffect.dashPathEffect(floatArrayOf(8f, 8f))
        if (fn == "cos") {
            drawLine(Theme.inkSoft, Offset(px, py), Offset(px, cy), strokeWidth = 2f, pathEffect = dash)
            drawLine(Theme.success, Offset(cx, cy), Offset(px, cy), strokeWidth = 4f)
        } else if (fn == "sin") {
            drawLine(Theme.inkSoft, Offset(px, py), Offset(cx, py), strokeWidth = 2f, pathEffect = dash)
            drawLine(Theme.success, Offset(cx, cy), Offset(cx, py), strokeWidth = 4f)
        }
    }
}

/** A vector at an angle, with the requested component picked out. */
@Composable
private fun VectorPlot(angleDeg: Double, component: String) {
    Canvas(Modifier.fillMaxWidth().height(180.dp)) {
        val cx = size.width / 2
        val cy = size.height / 2
        val r = minOf(cx, cy) * 0.78f

        drawLine(Theme.line, Offset(cx - r, cy), Offset(cx + r, cy), strokeWidth = 1.5f)
        drawLine(Theme.line, Offset(cx, cy - r), Offset(cx, cy + r), strokeWidth = 1.5f)

        val rad = angleDeg * PI / 180.0
        val px = cx + r * cos(rad).toFloat()
        val py = cy - r * sin(rad).toFloat()

        drawLine(Theme.gold300, Offset(cx, cy), Offset(px, py), strokeWidth = 4f)
        drawCircle(Theme.gold300, radius = 7f, center = Offset(px, py))

        val dash = PathEffect.dashPathEffect(floatArrayOf(8f, 8f))
        if (component == "x") {
            drawLine(Theme.inkSoft, Offset(px, py), Offset(px, cy), strokeWidth = 2f, pathEffect = dash)
            drawLine(Theme.success, Offset(cx, cy), Offset(px, cy), strokeWidth = 5f)
        } else {
            drawLine(Theme.inkSoft, Offset(px, py), Offset(cx, py), strokeWidth = 2f, pathEffect = dash)
            drawLine(Theme.success, Offset(cx, cy), Offset(cx, py), strokeWidth = 5f)
        }
    }
}
